src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java

```diff
@@ -0,0 +1,159 @@
+        Assertions.assertEquals(HttpStatus.BAD_REQUEST.value(), response.getStatus());
+    }
+
+    @Test
+    @WithMockUser
+    void requestPostWithValidBodyTest() throws Exception {
+        Mockito.when(productService.save(Mockito.any())).thenReturn(PRODUCT_DTO_ACTIVE);
+
+        MockHttpServletResponse response = mockMvc
+                .perform(
+                        MockMvcRequestBuilders
+                                .post(PRODUCT_URL)
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(productDTORequestJson.write(PRODUCT_DTO_ACTIVE).getJson())
+                )
+                .andReturn().getResponse();
+        String jsonExpected = productDTOResponseJson.write(PRODUCT_DTO_ACTIVE).getJson();
+
+        Assertions.assertEquals(HttpStatus.CREATED.value(), response.getStatus());
+        Assertions.assertEquals(JsonParser.parseString(jsonExpected), JsonParser.parseString(response.getContentAsString()));
+    }
+
+    @Test
+    @WithMockUser
+    void requestPutWithInvalidBodyTest() throws Exception {
+        MockHttpServletResponse response = mockMvc.perform(MockMvcRequestBuilders.put(PRODUCT_URL)).andReturn().getResponse();
+        Assertions.assertEquals(HttpStatus.BAD_REQUEST.value(), response.getStatus());
+    }
+
+    @Test
+    @WithMockUser
+    void requestPutWithValidBodyTest() throws Exception {
+        Mockito.when(productService.update(Mockito.any())).thenReturn(PRODUCT_DTO_ACTIVE);
+
+        MockHttpServletResponse response = mockMvc
+                .perform(
+                        MockMvcRequestBuilders
+                                .put(PRODUCT_URL)
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(productDTORequestJson.write(PRODUCT_DTO_ACTIVE).getJson())
+                )
+                .andReturn().getResponse();
+        String jsonExpected = productDTOResponseJson.write(PRODUCT_DTO_ACTIVE).getJson();
+
+        Assertions.assertEquals(HttpStatus.OK.value(), response.getStatus());
+        Assertions.assertEquals(JsonParser.parseString(jsonExpected), JsonParser.parseString(response.getContentAsString()));
+    }
+
+    @Test
+    @WithMockUser
+    void requestDeleteTest() throws Exception {
+        Mockito.doNothing().when(productService).delete(Mockito.anyLong());
+
+        MockHttpServletResponse response = mockMvc
+                .perform(
+                        MockMvcRequestBuilders
+                                .delete(PRODUCT_URL_WITH_ID, Mockito.anyLong())
+                                .contentType(MediaType.TEXT_PLAIN)
+                )
+                .andReturn().getResponse();
+
+        Assertions.assertEquals(HttpStatus.NO_CONTENT.value(), response.getStatus());
+    }
+
+    @Test
+    @WithMockUser
+    void requestFindTest() throws Exception {
```

Comment:
The case of results not found is tested in the resourceNotFoundTest() method. In this method I group all cases which some result is not found.

Reply:
I not do this cases because I mock return of the Service, and the Service methods I also tested.

The correct would be not to mock the return of the service and use the value it returns?

Reply:
No, you can keep them mocked. But you would mock to the values for such corner cases, then you can test if the intended result code and JSON response is being delivered as expected, and the controller is not messing up anything on the middle.

On a second note, there is also the possibility of not mocking the services there and implementing integration tests for the controller layer. Not sure how this would work for Spring, though.

Reply:
Hmm, got it.

Updated tests to use a fixed value.